### PR TITLE
Revamp granary and warehouse permissions

### PR DESCRIPTION
**Pull request by dvincent56**

### CODE REVIEW

**Comment by Ouaz** [comment]
> Revert caravanserai as red walker

- We did so because otherwise the caravanserai and work camp permission icons are the same (grey outfit walker)

- The sentry icon for the Supply Post permission (as currently) makes clearer that it's food that can be collected in this granary for the soldiers. The red outfit guy not so much.

- The monument architect instead of the work camp walker makes clearer it's materials that can be collected in this warehouse for the momument construction.

Imo, the grey and red outfit guys are too generic to depict a permission, whereas we can use other available figures, which are more iconic.

**Comment by Areldir** [comment]
Buildings have unique walkers for a reason. Using the wrong walker on the button will confuse players, who expect the walker depicted in the UI to be the one that the building actually sends. That is why I asked for the change 7cd7d68 to be made to this PR.

**Comment by Ouaz** [comment]
I don't understand. ^^

The Supply Post sends a Sentry figure (quartermaster) first to the granary, then he comes back with food bearers behind him. So the Sentry icon was more logic than the red guy (and clearer it's for the military food supply)

We changed the caravanserai walker to red in the game, and used it for the permission icon, to avoid having two times the grey guy. We could have changed his outfit to yellow or something else later.

Having **exactly** the buiding walker as the icon is not mandatory though.
The trade ship doesn't go itself to the building. ^^
And yet everybody understands that the button allows sea trade access for this building.
That would work the same with the monument architect: easy to understand it allows use of the materials stored in the warehouse, for monument construction.

But that's just my two cents, as always...

**Comment by crudelios** [comment]
This is why I prefer text to icons, but granted, it would be impossible to fit text here.

But we'll need at least some tooltips.

File: src/window/building/distribution.c [hunk 1]
```diff
@@ -194,41 +199,84 @@ static void draw_accept_none_button(int x, int y, int focused)
...
     }
-    image_buttons_draw(x, y, image_buttons_maintain, button);
+    image_buttons_draw(c->x_offset + 421, c->y_offset + 10, image_buttons_maintain, button);
 }
 
 static void draw_granary_permissions_buttons(int x, int y, int buttons)
-{
-    int offsets[] = { 96, 96, 132, 96 };
+{   
+    int images_permission[] = {
```
Comment (crudelios) [inline src/window/building/distribution.c:248]:
```suggestion
    static int images_permission[] = {
```

This way the image id‘s are only calculated once.

File: src/window/building/distribution.c [hunk 2]
```diff
@@ -194,41 +199,84 @@ static void draw_accept_none_button(int x, int y, int focused)
...
+    int image_offset_x, image_offset_y;
+
     for (int i = 0; i < buttons; i++) {
         int permission = granary_distribution_permissions_buttons[i].parameter1 - 1;
-        button_border_draw(x, y, 20, 20, data.permission_focus_button_id == i + 1 ? 1 : 0);
-        if (building_storage_get_permission(permission, building_get(data.building_id))) {
-            image_draw(assets_get_image_id("UI", "Allowed_Walker_Check"), x + 4, y + 4,
-                COLOR_MASK_NONE, SCALE_NONE);
+        int is_sea_trade_route = permission == BUILDING_STORAGE_PERMISSION_DOCK;
+        int permissionState = building_storage_get_permission(permission, building_get(data.building_id));
```
Comment (crudelios) [inline src/window/building/distribution.c:262]:
```suggestion
        int permission_state = building_storage_get_permission(permission, building_get(data.building_id));
```

File: src/window/building/distribution.c [hunk 3]
```diff
@@ -194,41 +199,84 @@ static void draw_accept_none_button(int x, int y, int focused)
...
+        image_offset_y = is_sea_trade_route ? 16 : 7;
+        
+        image_draw(images_permission[i], x + image_offset_x, y + image_offset_y, COLOR_MASK_NONE, SCALE_NONE);
+
+        if (!permissionState) {
+            image_draw(assets_get_image_id("UI", "Large_Widget_Cross"), x + 15, y + 15,
+            COLOR_MASK_NONE, SCALE_NONE);
+        }
+        
+        button_border_draw(x, y, 52, 52, data.permission_focus_button_id == i + 1 || !permissionState ? 1 : 0);
```
Comment (crudelios) [inline src/window/building/distribution.c:277]:
```suggestion
        button_border_draw(x, y, 52, 52, data.permission_focus_button_id == i + 1 || !permission_state);
```

This also works and is simpler to read.

**Review by crudelios — CHANGES_REQUESTED** [review]
Small changes needed, otherwise looks good. It already has the tooltips, so good job.

File: src/window/building/distribution.c (hunk 1 above)
Comment (dvincent56) [reply src/window/building/distribution.c:248]: not working 
Error: initializer element is not constantGCC

Comment (crudelios) [reply src/window/building/distribution.c:248]: Yeah makes sense now that I think about it.

**Comment by dvincent56** [comment]
@crudelios Do you have idea to fix warehouse and granary flags ?

**Comment by Areldir** [comment]
Speaking of the granary and warehouse flags, we will need new variants for the additional building permissions, right?

**Comment by dvincent56** [comment]
Yes probably.
Moreover, since you pushed new assets, flags are not working anymore.

**Comment by Ouaz** [comment]
> we will need new variants for the additional building permissions, right?

Ideally, yes.

But there's too many possible combinations now to make additional colored flags + icons a relevant thing.
As currently there's one unique flag for each possible combination with the 4 "core" permissions (Market, Supply Post / Armoury, Land Trade, Sea Trade).

One workaround would be too add different smaller flags for "optional" permissions  (Tavern, Caravanserai, Lighthouse, Work Camp), like small colored "pennants" hanging on the granary/warehouse roof.

perm_flags

**Comment by dvincent56** [comment]
@crudelios Can you take a look on this one. I don't know how to fix the granary / warehouse flags and maybe the problem come from another change ? @Areldir wants to remove some flags to keep only market / dock / caravan I think.

**Comment by crudelios** [comment]
Areldir just added some changes to the flags.

What's left to do?

**Comment by dvincent56** [comment]
@crudelios I pushed an update with separate buttons by priority as @ouaz suggest.
The last issue is get_warehouse_flag_image_id and get_granary_flag_image_id methods which are broken now and flags are not correct but I don't know if it due to my PR or not and I don't understand how these methods works

**Comment by crudelios** [comment]
Please check our discord group, we've been discussing the button groupings over there 🙂

**Pull request merged**